File: cpp/src/server/photos_handler/photo_results_location_accessor.cpp
/*
 WebServe
 Copyright 2018-2022 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

#include "photo_results_location_accessor.h"

#include "utils/string_helpers.h"

static const std::string kBuildLocationItemSeparator = "/";
static const std::string kLookupLocationItemSeparator = "/";

PhotoResultsLocationAccessor::PhotoResultsLocationAccessor()
{

}

PhotoResultsLocationAccessor::~PhotoResultsLocationAccessor()
{
	for (LocationHierarchyItem* pItem : m_aItems)
	{
		delete pItem;
	}
}

void PhotoResultsLocationAccessor::build(const std::vector<const PhotoItem*>& rawItems)
{
	// build a temporary cache of full path lookups from string hashes, so we can fairly quickly append the photo pointer
	// to the appropriate vectors without having to parse the item.

	std::map<HashValue, TempLocationHierarchyItemPtrs*> cachedFullLookup;

	for (const PhotoItem* pPhotoItem : rawItems)
	{
		const PhotoItem& photoItem = *pPhotoItem;

		if (!photoItem.getTimeTaken().isValid())
			continue;

		if (photoItem.getGeoLocationPath().isEmpty())
			continue;

		// first of all see if we've seen the full location value before, in which case we should
		// have a cache of the vector items
		HashValue fullLocationHash = photoItem.getGeoLocationPath().getHashValue();

		auto itFindFull = cachedFullLookup.find(fullLocationHash);

		if (itFindFull != cachedFullLookup.end())
		{
			// we have seen it before, so we can just append pointers

			const TempLocationHierarchyItemPtrs* pCachedVecPointers = itFindFull->second;
			for (unsigned int i = 0; i < TempLocationHierarchyItemPtrs::eTempLocationPointersNumber; i++)
			{
				LocationHierarchyItem* pVec = pCachedVecPointers->items[i];
				if (pVec)
				{
					pVec->photos.emplace_back(pPhotoItem);
				}
			}

			continue;
		}

		// otherwise, we haven't seen it before, so we need to break down the location hierarchy string into component items

		std::string locationString = photoItem.getGeoLocationPath().getString();
		StringHelpers::stripWhitespace(locationString);

		std::vector<std::string> locationComponents;
		StringHelpers::split(locationString, locationComponents, "/");

		// build up a temporary cache of pointers to the LocationHierarchyItems
		TempLocationHierarchyItemPtrs* pNewTempPointers = new TempLocationHierarchyItemPtrs();

		unsigned int level = 0;
		for (std::string& compStr : locationComponents)
		{
			StringHelpers::stripWhitespace(compStr);

			Hash hasher;
			hasher.addString(compStr);

			HashValue componentHash = hasher.getHash();

			if (level == 0)
			{
				auto itFindComponent = m_locationLookup.find(componentHash);

				if (itFindComponent == m_locationLookup.end())
				{
					// we didn't find it, so we need to create it
					LocationHierarchyItem* pNewItem = new LocationHierarchyItem(compStr);

					unsigned int newItemIndex = m_aItems.size();

					m_aItems.emplace_back(pNewItem);

					pNewTempPointers->items[0] = pNewItem;

					m_locationLookup.insert(std::pair<HashValue, unsigned int>(componentHash, newItemIndex));
					m_locationLookupAlphabetical.insert(std::pair<std::string, unsigned int>(compStr, newItemIndex));
				}
				else
				{
					// we've seen this first level item before
					unsigned int firstLevelComponentItemIndex = itFindComponent->second;
					LocationHierarchyItem* pFirstLevelItem = m_aItems[firstLevelComponentItemIndex];

					pNewTempPointers->items[0] = pFirstLevelItem;
				}
			}
			else
			{
				// do the remaining items

				// for the moment, clamp number of components we can handle...
				if (level >= TempLocationHierarchyItemPtrs::eTempLocationPointersNumber)
					break;

				LocationHierarchyItem* previousLevelItem = pNewTempPointers->items[level - 1];

				auto itFindComponent = previousLevelItem->subLocationLookup.find(componentHash);

				if (itFindComponent == previousLevelItem->subLocationLookup.end())
				{
					// we didn't find it, so create a new item...

					LocationHierarchyItem* pNewSubItem = new LocationHierarchyItem(compStr);

					unsigned int newItemIndex = m_aItems.size();

					m_aItems.push_back(pNewSubItem);

					pNewTempPointers->items[level] = pNewSubItem;

					previousLevelItem->subLocationLookup.insert(std::pair<HashValue, unsigned int>(componentHash, newItemIndex));
					previousLevelItem->subLocationLookupAlphabetical.insert(std::pair<std::string, unsigned int>(compStr, newItemIndex));
				}
				else
				{
					// we've seen this level already...
					unsigned int thisSubItemIndex = itFindComponent->second;
					LocationHierarchyItem* pThisSubItem = m_aItems[thisSubItemIndex];

					pNewTempPointers->items[level] = pThisSubItem;
				}
			}

			level++;
		}

		// so we now should have an array of the items to add the photo pointer to within tempPointers.items[]...

		// as we haven't seen this full item before, add it
		cachedFullLookup.insert(std::pair<HashValue, TempLocationHierarchyItemPtrs*>(fullLocationHash, pNewTempPointers));

		for (unsigned int i = 0; i < level; i++)
		{
			if (!pNewTempPointers->items[i])
				break;

			LocationHierarchyItem* pLocationItem = pNewTempPointers->items[i];
			pLocationItem->photos.emplace_back(pPhotoItem);
		}
	}
}

std::vector<std::string> PhotoResultsLocationAccessor::getSubLocationsForLocation(const std::string& locationPath) const
{
	std::vector<std::string> subLocations;

	if (locationPath.empty())
	{
		// we're at the root, so return the first level items...
		for (const auto& item : m_locationLookupAlphabetical)
		{
			subLocations.push_back(item.first);
		}
		return subLocations;
	}

	// otherwise, follow the components...

	std::vector<std::string> locationComponents;
	StringHelpers::split(locationPath, locationComponents, "/");
	for (std::string& compStr : locationComponents)
	{
		StringHelpers::stripWhitespace(compStr);
	}

	unsigned int itemIndex = 0;
	LocationHierarchyItem* pItem = nullptr;
	unsigned int componentCount = 0;
	for (const std::string& component : locationComponents)
	{
		if (componentCount++ == 0)
		{
			auto itFind = m_locationLookupAlphabetical.find(component);
			if (itFind == m_locationLookupAlphabetical.end())
				return subLocations;

			itemIndex = itFind->second;
		}
		else if (pItem)
		{
			auto itFind = pItem->subLocationLookupAlphabetical.find(component);
			if (itFind == pItem->subLocationLookupAlphabetical.end())
				return subLocations;

			itemIndex = itFind->second;
		}
		else
		{
			return subLocations;
		}

		pItem = m_aItems[itemIndex];
	}

	if (pItem)
	{
		// we want them to be alphabetical, so we iterate the map
		for (const auto& subItem : pItem->subLocationLookupAlphabetical)
		{
			subLocations.push_back(subItem.first);
		}
	}

	return subLocations;
}

const std::vector<const PhotoItem*>* PhotoResultsLocationAccessor::getPhotosForLocation(const std::string& locationPath) const
{
	if (locationPath.empty())
		return nullptr;

	std::vector<std::string> locationComponents;
	StringHelpers::split(locationPath, locationComponents, "/");
	for (std::string& compStr : locationComponents)
	{
		StringHelpers::stripWhitespace(compStr);
	}

	unsigned int itemIndex = 0;
	LocationHierarchyItem* pItem = nullptr;
	unsigned int componentCount = 0;
	for (const std::string& component : locationComponents)
	{
		if (componentCount++ == 0)
		{
			auto itFind = m_locationLookupAlphabetical.find(component);
			if (itFind == m_locationLookupAlphabetical.end())
				return nullptr;

			itemIndex = itFind->second;
		}
		else if (pItem)
		{
			auto itFind = pItem->subLocationLookupAlphabetical.find(component);
			if (itFind == pItem->subLocationLookupAlphabetical.end())
				return nullptr;

			itemIndex = itFind->second;
		}
		else
		{
			return nullptr;
		}

		pItem = m_aItems[itemIndex];
	}

	if (pItem)
	{
		return &pItem->photos;
	}

	return nullptr;
}

File: rs/src/photos_handler/photo_results_location_accessor.rs
/*
 WebServe (Rust port)
 Copyright 2021-2024 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

use std::collections::BTreeMap;

use std::sync::Arc;

use super::photo_item::PhotoItem;

#[derive(Clone, Debug)]
struct LocationHierarchyItem {
    name:                   String,
    description:            String,

    sub_location_lookup:    BTreeMap<String, u32>,
    photos:                 Vec<Arc<Box<PhotoItem>>>
}

impl LocationHierarchyItem {
    pub fn new(name: &str) -> LocationHierarchyItem {
        LocationHierarchyItem { name: name.to_string(), description: String::new(),
                                sub_location_lookup: BTreeMap::new(), photos: Vec::new() }
    }
}

#[derive(Clone, Debug)]
pub struct PhotoResultsLocationAccessor {
    items:                  Vec<LocationHierarchyItem>,

    location_lookup:        BTreeMap<String, u32>
}

impl PhotoResultsLocationAccessor {
    pub fn new() -> PhotoResultsLocationAccessor {
        PhotoResultsLocationAccessor { items: Vec::new(), location_lookup: BTreeMap::new() }
    }

    pub fn build(&mut self, photos: &Vec<Arc<Box<PhotoItem>>>) {
        for photo in photos {
            // TODO: do we care about this for locations?
            if photo.time_taken.is_none() {
                continue;
            }

            if photo.geo_location_path.is_empty() {
                continue;
            }

            let location_string = photo.geo_location_path.trim();

            let location_components: Vec<&str> = location_string.split('/').map(|x| x.trim()).collect();

            let mut new_temp_items : Vec<usize> = Vec::with_capacity(4);

            let mut level = 0;
            for comp_string in location_components {
                if level == 0 {
                    let find_comp = self.location_lookup.get(comp_string);
                    if let Some(comp) = find_comp {
                        // we've seen this first level item before...
                        let first_level_component_item_index = comp;

                        new_temp_items.push(*first_level_component_item_index as usize);
                    }
                    else {
                        // we didn't find it, so we need to create it...
                        let new_item_index = self.items.len();

                        self.items.push(LocationHierarchyItem::new(comp_string));

                        self.location_lookup.insert(comp_string.to_string(), new_item_index as u32);

                        new_temp_items.push(new_item_index);
                    }
                }
                else {
                    // do the remaining items...

                    let previous_level_item_index = new_temp_items[level - 1_usize];
                    let previous_level_item = &self.items[previous_level_item_index];
                    let find_comp = previous_level_item.sub_location_lookup.get(comp_string);

                    if let Some(comp) = find_comp {
                        // we've seen this level item before...
                        let this_sub_item_index = comp;

                        new_temp_items.push(*this_sub_item_index as usize);
                    }
                    else {
                        // we didn't find it, so we need to create it...
                        let new_item_index = self.items.len();

                        self.items.push(LocationHierarchyItem::new(comp_string));

                        let previous_level_item = &mut self.items[previous_level_item_index];
                        previous_level_item.sub_location_lookup.insert(comp_string.to_string(), new_item_index as u32);

                        new_temp_items.push(new_item_index);
                    }
                }

                level += 1;
            }

            // now we should have an array of the items to add the photo to...

            for ind in new_temp_items {
                let location_item = &mut self.items[ind];
                location_item.photos.push(photo.clone());
            }
        }
    }

    pub fn get_photos_for_location(&self, location_path: &str) -> Option<&Vec<Arc<Box<PhotoItem>>>> {
        if location_path.is_empty() {
            return None;
        }

        let location_components: Vec<&str> = location_path.split('/').map(|x| x.trim()).collect();

        let mut item: Option<&LocationHierarchyItem> = None;
        let mut component_count = 0;

        for component in location_components {
            if component_count == 0 {
                if let Some(index) = self.location_lookup.get(component) {
                    item = Some(&self.items[*index as usize]);
                }
                else {
                    return None;
                }
            }
            else if let Some(sub_item) = item {
                if let Some(index) = sub_item.sub_location_lookup.get(component) {
                    item = Some(&self.items[*index as usize]);
                }
            }
            else {
                return None;
            }

            component_count += 1;
        }

        if let Some(item_val) = item {
            return Some(&item_val.photos);
        }

        None
    }

    pub fn get_sub_locations_for_location(&self, location_path: &str) -> Vec<String> {
        let mut sub_locations = Vec::new();

        if location_path.is_empty() {
            // we're at the root, so return the first level items...
            for item in self.location_lookup.keys() {
                sub_locations.push(item.clone());
            }
            return sub_locations;
        }

        // otherwise, follow the components...
        let location_components: Vec<&str> = location_path.split('/').map(|x| x.trim()).collect();

        let mut item: Option<&LocationHierarchyItem> = None;
        let mut component_count = 0;

        for component in location_components {
            if component_count == 0 {
                if let Some(index) = self.location_lookup.get(component) {
                    item = Some(&self.items[*index as usize]);
                }
                else {
                    return sub_locations;
                }
            }
            else if let Some(sub_item) = item {
                if let Some(index) = sub_item.sub_location_lookup.get(component) {
                    item = Some(&self.items[*index as usize]);
                }
            }
            else {
                return sub_locations;
            }

            component_count += 1;
        }

        if let Some(item_val) = item {
            for item in item_val.sub_location_lookup.keys() {
                sub_locations.push(item.clone());
            }
        }

        sub_locations
    }
}
